docs/dynamic-component-loading.md

```diff
@@ -0,0 +1,117 @@
+https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/assembly-conflict-resolution.md).
+
+## Dynamic loading with dependencies
+We propose to add a new public API which would dynamically load a component with these properties:
+* Component is loaded in isolation from the app (and other components) so that potential collisions are not an issue
+* Component can use `.deps.json` to describe its dependencies. This includes the ability to describe additional NuGet packages, RID-specific and/or native dependencies
+* Component can chose to rely on the app for certain dependencies by not including them in its `.deps.json`
+* Optionally such component can be enabled for unloading
+
+Public API (early thinking):
+```csharp
+class Assembly
+{
+    public static Assembly LoadFileWithDependencies(string path);
+}
+```
+
+In its core this is similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json` and such. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
+
+```csharp
+class AssemblyLoadContext
+{
+    public static AssemblyLoadContext CreateForAssemblyWithDependencies(string assemblyPath);
+
+    public static AssemblyLoadContext CreateForAssemblyWithDependencies(
+        string assemblyPath,
+        AssemblyLoadContext fallbackContext,
+        bool enableUnloading);
+}
+```
+
+"Advanced" version which would return an ALC instance and not just the main assembly. Allows for additional changes to the `AssemblyLoadContext`, like registering an event handler to the `Resolving` event for example.  
+The second overload adds the ability to specify:
+* `fallbackContext` which is the `AssemblyLoadContext` to defer to when the assembly resolution is not possible in this current context. By default this is the `AssemblyLoadContext.Default` (the app's context). This allows for creating effectively parent-child relationships between the load contexts.
+* `enableUnloading` which will mark the newly create load context to track references and to trigger unload fo the load context and all the assemblies in it when possible.
+
+## High-level description of the implementation
+* Implement a new `AssemblyLoadContext` which will provide the isolation boundary and act as the "root" for the component. It can be enabled for unloading.  
+* The new load context is initialized by specifying the full path to the main assembly of the component to load.  
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
+* Parsing and understanding of the `.deps.json` will be performed by the same host components which do this for executable apps (so same behavior/quirks/bugs, very little code duplication).
```

Comment:
Can you describe the "host" components? Does custom hosting require providing an implementation of this mechanism?

Reply:
I guess the host-components here is the code in hostpolicy.dll that does `deps.json` parsing. If a custom host should call `LoadWithDependencies` then it must provide an implementation of this functionality. We should consider whether it is worth extracting the deps.json parsing functionality to a separate DLL to share with other hosting scenarios.